Fractional Frame Loops

daschuer opened the pull request:
This allows to set loop start and end somewhere in the middle of two frames. 
This can happen if yo play the track at other rates than 1. 
Before, the positions where rounded to a full frame, which changes the size of the original loop. 
After a great amount of loop cycles, these errors could be summed to a significant error. 

A unittest is in place to demonstrate the issue. 

Still open: Allow to store loops at fractional frames.

daschuer committed:
Added poition rounding functions to SampleUtil

daschuer committed:
Handle fractional playposition in readaheadmanager

daschuer committed:
No need to zero out buffer, because the zeros are not used

daschuer committed:
Handle overshoot in fractional frame loops. Renamed buffer to pOutput and removed local copy

daschuer committed:
Remove rounding when returning from m_pReadAheadManager

daschuer committed:
Unse SampleUtil functions in enginebufferrubberband

daschuer committed:
Remove commended code

daschuer committed:
Added Test for a FractionalFrameLoop

illuusio commented:
Hmm. I don't really get this? Even checking unit test how can do this on GUI?

daschuer commented:
This PR removes the rounding errors when the selected loop does not match an integer frame value of the source track.

If you play a track with half speed, Mixxx puts additional interpolated artificial samples between all source track samples. If you now set a loop in point at start frame 0 and a loop out point to such an artificial sample, you have set a loop withe a fractional frame length, e.g. 10.5. of the source track.

Looping to this loop 10 times at normal speed should produce 105 frames at the sound card and not only 100 which is the case without this PR.

daschuer commented:
@uklotzde: Do you have fun to put an eye on this? That would be great :-)

uklotzde commented on src/engine/readaheadmanager.cpp:
```diff
@@ -23,88 +25,119 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
         }
     }
 
+    int start_sample;
```
We should use SINT here (and also in SampleUtil) for integer calculation with sample/frame positions and sample/frame/channel counts.

uklotzde commented on src/engine/readaheadmanager.cpp:
```diff
@@ -23,88 +25,119 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
+    int start_sample;
     if (in_reverse) {
-        start_sample = m_iCurrentPosition - samples_needed;
+        start_sample = SampleUtil::roundPlayPosToFrameStart(
```
...and CachingReader already delivers the result samples in reverse order, so the "start" position should match this direction -> Inconsistent interface design of CachingReader.

daschuer committed:
Use SINT as number of samples in SampleUtil
Unfortunately noat all loops a vercorized wit SINT soe have to remain in int.
Tested wit g++ 4.8.4

daschuer committed:
CachingReader calculates the start sample internaly

daschuer commented:
Thank you for the comments. 
They are resolved. Is it merge-able now?

uklotzde commented on src/engine/cachingreader.cpp:
```diff
@@ -222,7 +222,16 @@ void CachingReader::process() {
 }
 
-SINT CachingReader::read(SINT sample, bool reverse, SINT numSamples, CSAMPLE* buffer) {
+SINT CachingReader::read(SINT start_sample, bool reverse, SINT numSamples, CSAMPLE* buffer) {
```
The parameters start_sample and numSample belong closer together. The parameter reverse controls how this chunk of samples is copied into the buffer. Order: startSample, numSamples, reverse, buffer.

I'm still unsure which sample should be denoted "start" or "first" in case of reverse directions?!? When switching from sample to frame positions I would recommend to add a new class that exactly defines how a consecutive number of (complete) sample frames is specified, i.e. which frame is denoted as the first/start frame. One member of this class would be the direction, e.g. enum class <SomeGenericPrefix: "Audio"?>Direction { FORWARD, REVERSE }. Even in the current code base all those reverse flags could be replaced step by step by such a type-safe enum class.

uklotzde replied:
If we revert back to the old version where the interpretation of start_sample already depends on value of reverse (see below) you could keep the ordering of the parameters.

uklotzde commented on src/engine/readaheadmanager.cpp:
```diff
@@ -23,88 +25,113 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
+            m_currentPosition = loop_target;
+            if (preloop_samples > 0) {
+                // we are up to one frame ahead of the loop trigger
+                double overshoot = preloop_samples - samplesToLoopTrigger;
```
DEBUG_ASSERT(preloop_samples >= samplesToLoopTrigger);
double overshoot = preloop_samples - samplesToLoopTrigger;
// we are up to one frame ahead of the loop trigger
DEBUG_ASSERT(overshoot < kNumChannels);

uklotzde commented on src/engine/readaheadmanager.cpp:
```diff
@@ -23,88 +25,113 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
+            int loop_read_position = SampleUtil::roundPlayPosToFrameStart(
+                    m_currentPosition + (in_reverse ? preloop_samples : -preloop_samples), kNumChannels);
 
             int looping_samples_read = m_pReader->read(
```
Here you still use the old way of passing the first sample to the reader. Maybe it's safer to revert what I told you and always specify the first sample (= anchor) depending on the direction. This feels more "correct" to me now that I look at the resulting code. Doing lots of conditional computations depending on reverse or not is more a deficiency of the ReadAheadManager and a missing abstraction for handling directional audio data (see above, we need to encapsulate all this into a class sometime)

uklotzde commented on src/engine/readaheadmanager.cpp:
```diff
@@ -150,10 +177,13 @@ void ReadAheadManager::hintReader(double dRate, HintVector* pHintList) {
     current_position.sample = in_reverse ?
-            m_iCurrentPosition - length_to_cache :
-            m_iCurrentPosition;
+            sample - length_to_cache :
```
And here again, the first sample depends on the direction. I didn't notice before, so please revert the API of CachingReader back to the old way where the value of first sample already depends on the direction.

uklotzde commented:
Exactly, startFrame should always point to the first frame that is actually 
copied into the first position of the buffer. This follows the definition of 
reverse iterators in C++, where begin() always points to the first element.

On 15.11.2016 16:17, Daniel Schürmann wrote:
>
> *@daschuer* commented on this pull request.
>
> ----------------------------------------------------------------------------
>
> In src/engine/readaheadmanager.cpp 
> <https://github.com/mixxxdj/mixxx/pull/986>:
>
> >
> -            int loop_read_position = m_iCurrentPosition +
> -                    (in_reverse ? preloop_samples : -preloop_samples);
> +            // start reading before the loop start point, to crossfade these samples
> +            // with the samples we need to the loop end
> +            int loop_read_position = SampleUtil::roundPlayPosToFrameStart(
> +                    m_currentPosition + (in_reverse ? preloop_samples : -preloop_samples), kNumChannels);
>   
>               int looping_samples_read = m_pReader->read(
>
> I will test it in depth, but for now it looks like it is correct now, and 
> was wrong before.
> I just wonder why our looping test have not catches this. :-/
>
> I have not yet understand how a wrapping class could be actually helpful here.
>
> Your proposed API reads nicely and I am about to do this change,but switch 
> to a frame interface:
>
> |CachingReader::read(SINT startFrame, SINT numFrames, bool reverse, 
> CSAMPLE* buffer)|
>
>     Dear CachingReader, please read from "startFrame", "numFrames" in
>     "reverse" direction into buffer.
>
> Inside CachingReader I think "forwardFrameIndex" should clarify the other 
> type of index well.
>
> Any thoughts or ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/986>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-asZ0B8agdWfqdT0v95YaAR4oQVnFks5q-c0SgaJpZM4JXn_7>.
>

daschuer commented:
Yes, there was actually a bug when crossfading loop in reverse. 
https://bugs.launchpad.net/mixxx/+bug/1642050
Luckily, it was fixed by accident here :-D

daschuer committed:
reorder CachingReader::read() parameter

daschuer commented:
Ok, I was about to prepare a commit that uses frames in the read() call. 
But than I notice that this only cleans the code inside read() but with a cost of dirty the calling code.
The issue is that we have the m_playposition in samples and changing this to frames would be a big deal, with a doubtful benefit. Finally I have reverted all this and consider this PR as finished. 

@uklotzde: Any more comments? Is this mergable now?

uklotzde commented:
I didn't mean to switch to frame calculations right now ;)

We should do this step by step, starting at the SoundSources which already 
have been modified. If you start somewhere in the middle you need lots of 
back and forth transformations between samples (legacy code) and frames 
(refactored code).

I did not check all the loop calculations in detail, but I was able to 
follow your explanations. If you tested the various cases thoroughly I 
consider this work mergeable.

On 17.11.2016 00:11, Daniel Schürmann wrote:
>
> Ok, I was about to prepare a commit that uses frames in the read() call.
> But than I notice that this only cleans the code inside read() but with a 
> cost of dirty the calling code.
> The issue is that we have the m_playposition in samples and changing this 
> to frames would be a big deal, with a doubtful benefit. Finally I have 
> reverted all this and consider this PR as finished.
>
> @uklotzde <https://github.com/uklotzde>: Any more comments? Is this 
> mergable now?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/986#issuecomment-261103093>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-ancqTlN1Ut4ezxBsxEV2iS9RQ-qvks5q-42ygaJpZM4JXn_7>.
>

daschuer commented:
Thank you for your review!
